src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +672,477 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+          ctx.foreign_read,
+          ctx.deadline,
+          ctx.bytes_left,
+          ctx.srv.memory(),
+          ctx.srv.memory_fetch_sem());
+
+        // A read_result needs to be returned for every partition. Hence, the
+        // function can't return before the first call to
+        // `fetch_ntps_in_parallel`. However, if we weren't able to read the
+        // last_visible_index for a partition before the first call to
+        // `fetch_ntps_in_parallel` then we need to return with an error for
+        // that partition.
+        for (auto i : errored_partitions) {
+            results[i] = read_result(error_code::not_leader_for_partition);
+            results[i].partition = ctx.requests[i].ktp().get_partition();
+        }
+        errored_partitions.clear();
+
+        for (size_t i = 0; i < results.size(); i++) {
+            const auto& r = results[i];
+            total_size += r.data_size_bytes();
+            if (r.error != error_code::none) {
+                errored_partitions.push_back(i);
+            }
+        }
+
+        if (
+          total_size > ctx.min_bytes || !errored_partitions.empty()
+          || ctx.as.abort_requested()) {
+            co_return worker_result{
+              .read_results = std::move(results),
+              .errored_partitions = std::move(errored_partitions),
+              .total_size = total_size,
+            };
+        }
+
+        // If the fast-path doesn't return there isn't enough data so
+        // we'll have to wait for the last_visible_index of partitions to
+        // change.
+        std::vector<ss::future<>> hwm_waiters;
+        hwm_waiters.reserve(ctx.requests.size());
+        auto available_hwm_waiters = ss::make_lw_shared(std::vector<size_t>{});
+
+        // An abort source is needed to clear any waiters from the
+        // `raft::offset_monitor` after the worker returns.
+        ss::abort_source hwm_waiters_as;
+        auto hwm_as_sub = _as.subscribe(
+          [&hwm_waiters_as]() noexcept { hwm_waiters_as.request_abort(); });
+        auto waiter_cleanup = ss::defer(
+          [&hwm_waiters_as] { hwm_waiters_as.request_abort(); });
+
+        for (size_t i = 0; i < ctx.requests.size(); i++) {
+            auto wait_opt = register_hwm_wait(
+              ctx,
+              ctx.requests[i].ktp(),
+              model::next_offset(last_visible_indexes[i]),
+              hwm_waiters_as);
+
+            if (!wait_opt) {
+                results[i] = read_result(error_code::not_leader_for_partition);
+                results[i].partition = ctx.requests[i].ktp().get_partition();
+                co_return worker_result{
+                  .read_results = std::move(results),
+                  .errored_partitions = {i},
+                  .total_size = total_size,
+                };
+            }
+            auto waiter = wait_opt.value().finally([available_hwm_waiters, i] {
+                available_hwm_waiters->push_back(i);
```

Comment:
So picking up from the discussion from yesterday. Is this optimization even worth it if it doesn't properly work?

Reply:
I does work as it is today. The only thing that didn't work is my expectation that if `i` is in `available_hwm_waiters` then `hwm_waiters[i].available` is `true`. There are some cases where this doesn't hold. Though its completely safe to replace `hwm_waiters[i]` with another future even if its not available.

Reply:
Right ok, sounds good then.